firedrake/constant.py

```diff
@@ -49,43 +53,56 @@ class Constant(ufl.Coefficient, ConstantMixin):
        :class:`~ufl.form.Form` on its own you need to pass a
        :func:`~.Mesh` as the domain argument.
     """
+    _globalcount = 0
+
+    def __new__(cls, value, domain=None):
+        if domain:
+            # Avoid circular import
+            from firedrake.function import Function
+            from firedrake.functionspace import FunctionSpace
+            import warnings
+            warnings.warn("Constants with a domain defined are functions in the Real space")
+
+            dat, rank, shape = _create_dat(op2.Global, value, domain._comm)
+            dat.zero()
 
-    def __new__(cls, *args, **kwargs):
-        # Hack to avoid hitting `ufl.Coefficient.__new__` which may perform operations
-        # meant for coefficients and not constants (e.g. check if the function space is dual or not)
-        # This is a consequence of firedrake.Constant inheriting from ufl.Coefficient instead of ufl.Constant.
-        return object.__new__(cls)
+            domain = ufl.as_domain(domain)
+            cell = domain.ufl_cell()
+            if rank == 0:
+                element = ufl.FiniteElement("R", cell, 0)
+            elif rank == 1:
+                element = ufl.VectorElement("R", cell, 0, shape[0])
+            else:
+                element = ufl.TensorElement("R", cell, 0, shape=shape)
+
+            R = FunctionSpace(domain, element, name="firedrake.Constant")
+            return Function(R, val=dat).assign(value)
```

Comment:
Same question about `value`.

Reply:
The assign needs to happen for annotation to occur. We can get away with not zeroing the `dat`, but we need to assign to create an assign block.